core/types/subtyping.cc

```diff
@@ -10,6 +10,37 @@ namespace sorbet::core {
 
 using namespace std;
 
+namespace {
+bool compositeTypeDeepRefEqual(const OrType &o1, const OrType &o2);
+bool compositeTypeDeepRefEqual(const AndType &a1, const AndType &a2);
+bool compositeTypeDeepRefEqualHelper(const TypePtr &t1, const TypePtr &t2) {
+    if (t1 == t2) {
+        return true;
+    }
+    if (t1.tag() != t2.tag()) {
+        return false;
+    }
+    // t1 and t2 are the same kind of type.
+    if (isa_type<OrType>(t1)) {
+        return compositeTypeDeepRefEqual(cast_type_nonnull<OrType>(t1), cast_type_nonnull<OrType>(t2));
+    }
+    if (isa_type<AndType>(t1)) {
+        return compositeTypeDeepRefEqual(cast_type_nonnull<AndType>(t1), cast_type_nonnull<AndType>(t2));
+    }
+    return false;
+}
+
+// Returns 'true' if the tree of types stemming from this AndType are referentially equal.
+bool compositeTypeDeepRefEqual(const AndType &a1, const AndType &a2) {
+    return compositeTypeDeepRefEqualHelper(a1.left, a2.left) && compositeTypeDeepRefEqualHelper(a1.right, a2.right);
```

Comment:
[Hi, I'm a new (2 days in) contributor, and curious about this, Tuples, and Shapes.]

How do we know that we don't need to also check `compositeTypeDeepRefEqualHelper(a1.left, a2.right) && compositeTypeDeepRefEqualHelper(a1.right, a2.left);`? Or in general, they could be structured differently:
```
t1: (and (and a b) (and c d))
t2:(and a (and b (and c d)))
```

Reply:
👋 Hey there! That's a great question.

For performance reasons, `lub` is best-effort -- so it doesn't always return a least upper bound. This is _especially_ true for shape types, which are not well supported in Sorbet (they are [expeirmental](https://sorbet.org/docs/shapes) for a reason!) and which are expensive to lub precisely.

As you expect, this code will not work if left and right are flipped relative to one another in the second type. We could handle this robustly by also checking left against right and right against left, but then you have an exponential blowout as that gets applied recursively!

In the cases I was looking at, this particular compromise worked well enough to simplify the types I was seeing and improve a real performance problem on a complex piece of code.

Reply:
(To be completely fair, I haven't benchmarked the exponential case. It might be neutral. I may try it after this lands.)